lib/Pipeline/Containers.cpp

```diff
@@ -0,0 +1,104 @@
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/Support/Error.h"
+
+#include "revng/Pipeline/ContainerSet.h"
+#include "revng/Pipeline/Errors.h"
+
+using namespace Pipeline;
+using namespace llvm;
+using namespace std;
+
+ContainerSet ContainerSet::cloneFiltered(const ContainersStatus &Targets) {
+  ContainerSet ToReturn;
+  for (const auto &Pair : Content) {
+    const auto &ContainerName = Pair.first();
+    const auto &Container = Pair.second;
+
+    auto ExtractedNames = Targets.contains(ContainerName) ?
+                            Targets.at(ContainerName) :
+                            TargetsList();
+
+    auto Cloned = Container != nullptr ?
+                    Container->cloneFiltered(ExtractedNames) :
+                    nullptr;
+
+    ToReturn.add(ContainerName, *Factories[Pair.first()], move(Cloned));
+  }
+  revng_assert(ToReturn.Content.size() == Content.size());
+  return ToReturn;
+}
+
+bool ContainerSet::contains(const Target &Target) const {
+  return llvm::any_of(Content, [&Target](const auto &Container) {
+    return Container.second->enumerate().contains(Target);
+  });
+}
+
+Error ContainerSet::remove(const ContainersStatus &ToRemove) {
+  for (const auto &Target : ToRemove) {
+    const auto &ContainerName = Target.first();
+    const auto &NamesToRemove = Target.second;
+    if (not contains(ContainerName))
+      continue;
+    if (auto Ok = at(ContainerName).remove(NamesToRemove); not Ok)
+      return make_error<UnknownTarget>(NamesToRemove, ContainerName);
+  }
+  return Error::success();
+}
+
+void ContainerSet::intersect(ContainersStatus &ToIntersect) const {
```

Comment:
This method is declared in `ContainerSet.h` but defined in `Containers.cpp`. Either merge the headers or create `ContainerSet.cpp`.